File: Online-judge/leetcode/convert-sorted-list-to-binary-search-tree.py
'''
/**
 * @author Md.Shohag <shohag.fks@gmail.com>
 * Problem : https://leetcode.com/problems/convert-sorted-list-to-binary-search-tree/description/
 */

/**
 * @brief 
 * Time complexity : O(n)
 * Space complexity: O(logn)
 */
'''

from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
        
class Solution:
    def getLinkedListSize(self, head: Optional[ListNode]) -> int:
        size = 0
        ptr = head
        while ptr:
            ptr = ptr.next
            size += 1
        return size

    def sortedListToBST(self, head: Optional[ListNode]) -> Optional[TreeNode]:
        size: int = self.getLinkedListSize(head)

        def listToBST(left, right):
            nonlocal head
            if left > right:
                return None
            mid = left + (right - left) // 2
            leftNode = listToBST(left, mid - 1)

            node = TreeNode(head.val)
            node.left = leftNode
            head = head.next

            node.right = listToBST(mid + 1, right)
            return node
        return listToBST(0, size - 1)

        

File: Online-judge/leetcode/convert-sorted-list-to-binary-search-tree.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * @author Md.Shohag <shohag.fks@gmail.com>
 * Problem : https://leetcode.com/problems/convert-sorted-list-to-binary-search-tree/description/
 */

/**
 * @brief 
 * Time complexity : O(n)
 * Space complexity: O(n)
 */
 
 
 struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
 };

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 };

class Solution {
private:
    vector<int> list;
    
    void linkListToList(ListNode* head) {
        while(head) {
            list.push_back(head->val);
            head = head->next;
        }
    }

    TreeNode* listToTreeNode(int left, int right) {
        if (left > right) return nullptr;
        int mid = left + (right - left) / 2;
        TreeNode* node = new TreeNode(list[mid]);

        node->left = listToTreeNode(left, mid - 1);
        node->right = listToTreeNode(mid + 1, right);
        return node;
    }

public:
    TreeNode* sortedListToBST(ListNode* head) {
        linkListToList(head);
        return listToTreeNode(0, list.size() - 1);
    }
};